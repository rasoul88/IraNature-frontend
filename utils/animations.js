import { keyframes } from "styled-components";

export const moveInLeft = keyframes`
    0% {
        opacity: 0;
        transform: translateX(-10rem);
    }

    80% {
        transform: translateX(1rem);
    }

    100% {
        opacity: 1;
        transform: translateX(0);
    }
`;

export const moveInRight = keyframes`
    0% {
        opacity: 0;
        transform: translateX(10rem);
    }

    80% {
        transform: translateX(-1rem);
    }

    100% {
        opacity: 1;
        transform: translateX(0);
    }
`;

export const moveInBottom = keyframes`
    0% {
        opacity: 0;
        transform: translateY(3rem);
    }

    100% {
        opacity: 1;
        transform: translateY(0);
    }
`;

export const Shake = keyframes`
    0% {
        transform: translateX(-.3rem);
    }

    100% {
        transform: translateX(.3rem);
    }
`;

export const pageInRight = keyframes`
    0% {
        opacity: 0;
        transform: translateX(200%);
    }

    80% {
        transform: translateX(-10rem);
    }

    100% {
        opacity: 1;
        transform: translateX(0);
    }
`;
