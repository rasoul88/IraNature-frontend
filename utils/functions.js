import { toast } from "react-toastify";
import Spin from "../public/assets/icons/Spin.svg";

export const hexTorgba = (color, alpha) => {
  if (alpha === undefined) alpha = 1;
  const r = parseInt(color.substr(1, 2), 16);
  const g = parseInt(color.substr(3, 2), 16);
  const b = parseInt(color.substr(5, 2), 16);
  return `rgba(${r}, ${g}, ${b}, ${alpha})`;
};

export const createGradientbackground = (from, to) => {
  return `linear-gradient(to right bottom, ${from},${to})`;
};

export const createGradientText = (from, to) => {
  return `linear-gradient(to right bottom,${hexTorgba(from, 0.85)}, 
  ${hexTorgba(to, 0.85)})`;
};

export const showToast = (mode, message, onClose) => {
  toast.dismiss();
  toast[mode](message, {
    autoClose: 5000,
    theme: "colored",
    // onOpen: (props) => console.log("i opened"),
    onClose: onClose,
    style: {
      fontFamily: "inherit",
    },
    position: toast.POSITION.BOTTOM_CENTER,
  });
};

export const showLoadingToast = (message) => {
  toast.dismiss();
  toast.info(message, {
    autoClose: false,
    style: {
      fontFamily: "inherit",
      boxShadow: "0 1px 10px 0 rgb(0 0 0 / 30%), 0 2px 15px 0 rgb(0 0 0 / 25%)",
      color: "#333",
    },
    position: toast.POSITION.BOTTOM_CENTER,
    icon: <Spin style={{ transform: "scale(1.5)", fill: "#666" }} />,
  });
};

export const objectDateToString = (objDate) => {
  return `${objDate.year}/${objDate.month < 10 ? "0" : ""}${objDate.month}/${
    objDate.day < 10 ? "0" : ""
  }${objDate.day}`;
};

// export const stringDateToObject = (stringDate) => {
//   const arrDate = stringDate.split("/");
//   return { year: arrDate[0], month: arrDate[1], day: arrDate[2] };
// };
