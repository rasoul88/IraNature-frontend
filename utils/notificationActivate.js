import { post } from "axios";

const urlBase64ToUint8Array = (base64String) => {
  var padding = "=".repeat((4 - (base64String.length % 4)) % 4);
  var base64 = (base64String + padding).replace(/\-/g, "+").replace(/_/g, "/");

  var rawData = window.atob(base64);
  var outputArray = new Uint8Array(rawData.length);

  for (var i = 0; i < rawData.length; ++i) {
    outputArray[i] = rawData.charCodeAt(i);
  }
  return outputArray;
};

const configurePushSub = () => {
  if (!("serviceWorker" in navigator)) {
    return;
  }

  var reg;
  navigator.serviceWorker.ready
    .then(function (swreg) {
      reg = swreg;
      return swreg.pushManager.getSubscription();
    })
    .then(function (sub) {
      if (sub === null) {
        // Create a new subscription
        var vapidPublicKey =
          "BKMOXd1ARsXXz97HxAFb6tU2BmVdhDJkvpscv77UWadgtLidzKZUxHMloFHdOx18kuYxHkYUTHGzbW0I_cwwAyE";
        var convertedVapidPublicKey = urlBase64ToUint8Array(vapidPublicKey);
        return reg.pushManager.subscribe({
          userVisibleOnly: true,
          applicationServerKey: convertedVapidPublicKey,
        });
      } else {
      }
    })
    .then((newSub) => {
      if (newSub) return post("/subscriptions", newSub);
    })
    .then((res) => {
      console.log("new subscription", res);
    })
    .catch(function (err) {
      console.log(err);
    });
};

export const askForNotificationPermission = () => {
  Notification.requestPermission((result) => {
    if (result !== "granted") {
    } else {
      configurePushSub();
    }
  });
};
