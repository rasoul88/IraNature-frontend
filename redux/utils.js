import { objectDateToString } from "../utils/functions";
export const createQuery = (filterItems, dataLimits) => {
  let query = "";
  for (let item in filterItems) {
    if (item === "name" && filterItems[item].length > 0)
      query += `name[regex]=${filterItems[item]}&`;

    if (item === "days") {
      if (filterItems[item][0] > dataLimits[item].min)
        query += `duration[gte]=${filterItems[item][0]}&`;
      if (filterItems[item][1] < dataLimits[item].max)
        query += `duration[lte]=${filterItems[item][1]}&`;
    }

    if (item === "price") {
      if (filterItems[item][0] > dataLimits[item].min)
        query += `price[gte]=${filterItems[item][0]}&`;
      if (filterItems[item][1] < dataLimits[item].max)
        query += `price[lte]=${filterItems[item][1]}&`;
    }

    if (
      item === "dateRange" &&
      filterItems[item].from &&
      filterItems[item].to
    ) {
      query += `startDate[gte]=${objectDateToString(filterItems[item].from)}&`;
      query += `startDate[lte]=${objectDateToString(filterItems[item].to)}&`;
    }

    if (item === "startLocation" && filterItems[item] && filterItems[item][0]) {
      query += `startLocation=${filterItems[item][0].name}&`;
    }

    if (
      item === "difficulty" &&
      (filterItems[item].length === 1 || filterItems[item].length === 2)
    ) {
      filterItems[item].forEach((el) => (query += `difficulty=${el}&`));
    }

    if (
      item === "maxParticipants" &&
      filterItems[item][0] < dataLimits[item].max
    ) {
      query += `maxGroupSize[lte]=${filterItems[item][0]}&`;
    }

    if (item === "page") {
      query += `page=${filterItems[item]}&`;
    }
  }

  query = query.slice(0, query.length - 1);

  return query;
};
