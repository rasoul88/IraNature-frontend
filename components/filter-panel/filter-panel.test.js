import { shallow } from "enzyme";
import toJson from "enzyme-to-json";
import { findByTestAttribute, storeFactory } from "../../test/utils";
import FilterPanel, { UnconnectedFilterPanel } from "./filter-panel.component";

const setup = () => {
  const initialState = {
    tours: {
      panelStatus: false,
      filterItems: {
        name: "",
        days: [1, 20],
        price: [200000, 4000000],
        cities: null,
        dateRange: {
          from: null,
          to: null,
        },
        maxParticipants: [40],
        difficulty: [],
        page: 1,
      },
      dataLimits: {
        days: { min: 1, max: 20, step: 1 },
        price: { min: 200000, max: 4000000, step: 50000 },
        maxParticipants: { min: 5, max: 40, step: 1 },
      },
      activeCities: [
        { name: "تهران", id: 1 },
        { name: "اصفهان", id: 2 },
        { name: "تبریز", id: 3 },
        { name: "کرج", id: 4 },
        { name: "شیراز", id: 5 },
      ],
    },
  };
  const store = storeFactory(initialState);
  return shallow(<FilterPanel store={store} />)
    .dive()
    .dive();
};

test("renders filterPanel correctly", () => {
  const wrapper = setup();
  const component = findByTestAttribute(wrapper, "component-filter-panel");
  expect(component.length).toBe(1);
  expect(toJson(wrapper)).toMatchSnapshot();
});

describe("test action calls on filter item changes", () => {
  let wrapper;
  const mockSetFilterItem = jest.fn();
  const mocktoggleFilterPanel = jest.fn();

  beforeEach(() => {
    mockSetFilterItem.mockClear();
    mocktoggleFilterPanel.mockClear();
    const props = {
      panelStatus: false,
      filterItems: {
        name: "",
        days: [1, 20],
        price: [200000, 4000000],
        cities: null,
        dateRange: {
          from: null,
          to: null,
        },
        maxParticipants: [40],
        difficulty: [],
        page: 1,
      },
      dataLimits: {
        days: { min: 1, max: 20, step: 1 },
        price: { min: 200000, max: 4000000, step: 50000 },
        maxParticipants: { min: 5, max: 40, step: 1 },
      },
      activeCities: [
        { name: "تهران", id: 1 },
        { name: "اصفهان", id: 2 },
        { name: "تبریز", id: 3 },
        { name: "کرج", id: 4 },
        { name: "شیراز", id: 5 },
      ],
      toggleFilterPanel: mocktoggleFilterPanel,
      setFilterItem: mockSetFilterItem,
    };

    wrapper = shallow(<UnconnectedFilterPanel {...props} />);
  });

  test("test toggleFilterPanel button", () => {
    const toggleFilterPanelButton = findByTestAttribute(
      wrapper,
      "toggle-filter-panel-button"
    );
    expect(toggleFilterPanelButton.length).toBe(1);

    toggleFilterPanelButton.simulate("click");
    expect(mocktoggleFilterPanel).toHaveBeenCalled();
  });

  test("test name input", () => {
    const inputComponent = findByTestAttribute(wrapper, "name-input");
    expect(inputComponent.length).toBe(1);
    inputComponent.simulate("blur", { target: { value: "see explorer" } });
    expect(mockSetFilterItem).toHaveBeenCalledWith("name", "see explorer");
  });

  test("test price range slider", () => {
    const priceRangeSliderComponent = findByTestAttribute(
      wrapper,
      "price-range-slider"
    );
    expect(priceRangeSliderComponent.length).toBe(1);

    const simulateRange = [1000000, 3000000];
    priceRangeSliderComponent.simulate("change", simulateRange);
    expect(mockSetFilterItem).toHaveBeenCalledWith("price", simulateRange);
  });

  test("test days range slider", () => {
    const daysRangeSliderComponent = findByTestAttribute(
      wrapper,
      "days-range-slider"
    );
    expect(daysRangeSliderComponent.length).toBe(1);
    const simulateRange = [3, 7];
    daysRangeSliderComponent.simulate("change", simulateRange);
    expect(mockSetFilterItem).toHaveBeenCalledWith("days", simulateRange);
  });

  test("test dateRange date picker", () => {
    const dateRangeDatePickerComponent = findByTestAttribute(
      wrapper,
      "dateRange-date-picker"
    );
    expect(dateRangeDatePickerComponent.length).toBe(1);

    const simulateDate = {
      from: { day: 1, month: 8, year: 1400 },
      to: { day: 12, month: 8, year: 1400 },
    };
    dateRangeDatePickerComponent.simulate("change", simulateDate);
    expect(mockSetFilterItem).toHaveBeenCalledWith("dateRange", simulateDate);
  });

  test("test cities input selector", () => {
    const citiesSelectorComponent = findByTestAttribute(
      wrapper,
      "cities-multiselector"
    );
    expect(citiesSelectorComponent.length).toBe(1);
    const simulateCities = ["اصفهان", "تهران"];
    citiesSelectorComponent.simulate("select", simulateCities);
    expect(mockSetFilterItem).toHaveBeenCalledWith("cities", simulateCities);
  });

  test("test difficulty checkbox", () => {
    const difficultyCheckboxComponent = findByTestAttribute(
      wrapper,
      "difficulty-checkbox"
    );
    expect(difficultyCheckboxComponent.length).toBe(1);

    const simulateDifficulty = ["متوسط", "آسان"];
    difficultyCheckboxComponent.simulate("change", simulateDifficulty);
    expect(mockSetFilterItem).toHaveBeenCalledWith(
      "difficulty",
      simulateDifficulty
    );
  });

  test("test maxParticipants range slider", () => {
    const maxParticipantsRangeSliderComponent = findByTestAttribute(
      wrapper,
      "maxParticipants-range-slider"
    );
    expect(maxParticipantsRangeSliderComponent.length).toBe(1);
    const simulateRange = [17];
    maxParticipantsRangeSliderComponent.simulate("change", simulateRange);
    expect(mockSetFilterItem).toHaveBeenCalledWith(
      "maxParticipants",
      simulateRange
    );
  });
});
