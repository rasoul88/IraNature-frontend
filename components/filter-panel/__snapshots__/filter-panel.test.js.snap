// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders filterPanel correctly 1`] = `
<filter-panelstyles__PanelContainer
  data-test="component-filter-panel"
  isOpen={false}
>
  <filter-panelstyles__FilterIconContainer
    data-test="toggle-filter-panel-button"
    onClick={[Function]}
  >
    <test-file-stub />
  </filter-panelstyles__FilterIconContainer>
  <filter-panelstyles__ContentContainer
    isOpen={false}
  >
    <filter-panelstyles__PanelHeader>
      <p>
        4
      </p>
      <h4>
        فیلترها
      </h4>
    </filter-panelstyles__PanelHeader>
    <filter-panelstyles__FilterItemContainer>
      <filter-panelstyles__ItemHeaderContainer>
        <h5>
          نام 
        </h5>
      </filter-panelstyles__ItemHeaderContainer>
      <CustomInput
        data-test="name-input"
        onBlur={[Function]}
        placeholder="نام تور را وارد کنید"
        type="text"
      />
    </filter-panelstyles__FilterItemContainer>
    <filter-panelstyles__FilterItemContainer>
      <filter-panelstyles__ItemHeaderContainer>
        <h5>
          محدوده قیمت
        </h5>
        <p>
          از 
          200,000
           تا 
          4,000,000
        </p>
      </filter-panelstyles__ItemHeaderContainer>
      <CustomRangeSlider
        data-test="price-range-slider"
        max={4000000}
        min={200000}
        onChange={[Function]}
        step={50000}
        valueLabelDisplay="off"
        values={
          Array [
            200000,
            4000000,
          ]
        }
      />
    </filter-panelstyles__FilterItemContainer>
    <filter-panelstyles__FilterItemContainer>
      <filter-panelstyles__ItemHeaderContainer>
        <h5>
          مدت 
        </h5>
        <p>
          از 
          1
           تا 
          20
           روز
        </p>
      </filter-panelstyles__ItemHeaderContainer>
      <CustomRangeSlider
        data-test="days-range-slider"
        max={20}
        min={1}
        onChange={[Function]}
        step={1}
        valueLabelDisplay="on"
        values={
          Array [
            1,
            20,
          ]
        }
      />
    </filter-panelstyles__FilterItemContainer>
    <filter-panelstyles__FilterItemContainer>
      <filter-panelstyles__ItemHeaderContainer>
        <h5>
          محدوده زمانی شروع 
        </h5>
      </filter-panelstyles__ItemHeaderContainer>
      <CustomDatePicker
        data-test="dateRange-date-picker"
        onChange={[Function]}
        selectedRange={
          Object {
            "from": null,
            "to": null,
          }
        }
      />
    </filter-panelstyles__FilterItemContainer>
    <filter-panelstyles__FilterItemContainer>
      <filter-panelstyles__ItemHeaderContainer>
        <h5>
          شهر مبدا 
        </h5>
      </filter-panelstyles__ItemHeaderContainer>
      <Multiselect
        avoidHighlightFirstOption={false}
        caseSensitiveSearch={false}
        closeIcon="circle2"
        closeOnSelect={true}
        customCloseIcon=""
        data-test="cities-multiselector"
        disablePreSelectedValues={false}
        displayValue="name"
        emptyRecordMsg="در حال حاضر شهر دیگری برای انتخاب وجود ندارد"
        groupBy=""
        hidePlaceholder={false}
        id="multiselect-react-dropdown"
        isObject={true}
        keepSearchTerm={false}
        onRemove={[Function]}
        onSelect={[Function]}
        options={
          Array [
            Object {
              "id": 1,
              "name": "تهران",
            },
            Object {
              "id": 2,
              "name": "اصفهان",
            },
            Object {
              "id": 3,
              "name": "تبریز",
            },
            Object {
              "id": 4,
              "name": "کرج",
            },
            Object {
              "id": 5,
              "name": "شیراز",
            },
          ]
        }
        placeholder="انتخاب"
        selectedValues={null}
        selectionLimit={-1}
        showArrow={true}
        showCheckbox={false}
        singleSelect={false}
        style={
          Object {
            "inputField": Object {
              "fontFamily": "Iransans",
              "marginRight": "2rem",
              "textAlign": "right",
              "width": "5rem",
            },
            "optionContainer": Object {
              "fontSize": "14px",
              "textAlign": "right",
            },
            "searchBox": Object {
              "fontSize": "1.4rem",
              "minHeight": "4.6rem",
              "textAlign": "right",
            },
          }
        }
      />
    </filter-panelstyles__FilterItemContainer>
    <filter-panelstyles__FilterItemContainer>
      <filter-panelstyles__ItemHeaderContainer>
        <h5>
          درجه سختی
        </h5>
      </filter-panelstyles__ItemHeaderContainer>
      <CustomCheckbox
        data-test="difficulty-checkbox"
        onChange={[Function]}
        selectedValues={Array []}
      >
        <custom-check-boxstyles__CheckboxItem
          key="hard"
        >
          سخت
        </custom-check-boxstyles__CheckboxItem>
        <custom-check-boxstyles__CheckboxItem
          key="medium"
        >
          متوسط
        </custom-check-boxstyles__CheckboxItem>
        <custom-check-boxstyles__CheckboxItem
          key="easy"
        >
          آسان
        </custom-check-boxstyles__CheckboxItem>
      </CustomCheckbox>
    </filter-panelstyles__FilterItemContainer>
    <filter-panelstyles__FilterItemContainer>
      <filter-panelstyles__ItemHeaderContainer>
        <h5>
          حداکثر تعداد شرکت کننده
        </h5>
        <p>
          تا 
          40
           نفر
        </p>
      </filter-panelstyles__ItemHeaderContainer>
      <CustomRangeSlider
        data-test="maxParticipants-range-slider"
        max={40}
        min={5}
        onChange={[Function]}
        valueLabelDisplay="on"
        values={
          Array [
            40,
          ]
        }
      />
    </filter-panelstyles__FilterItemContainer>
    <CustomButton
      backgroundColor="#1976D2"
      color="white"
      style={
        Object {
          "margin": "2rem 0",
        }
      }
    >
      اعمال فیلتر ها
    </CustomButton>
  </filter-panelstyles__ContentContainer>
</filter-panelstyles__PanelContainer>
`;
